### PR TITLE
Rebalancing

### PR DESCRIPTION
**Pull request by Suyog007**
Rebalancing contract SCORE for the case when the oracle price is greater than dex pool price.

### CODE REVIEW

File: rebalancing/rebalancing.py [hunk 1]
```diff
@@ -0,0 +1,311 @@
...
+    @external(readonly=True)
+    def getRebalancingStatus(self) -> list:
+        """
+        Checks the Rebalancing status of the pool.
+        """
+        rebalancing_direction = 0
+        self.bnUSD_score = self.create_interface_score(self._bnUSD.get(), bnUSDTokenInterface)
+        self.dex_score = self.create_interface_score(self._dex.get(), dexTokenInterface)
+        self.loans_score = self.create_interface_score(self._loans.get(), loansTokenInterface)
+        price = self.bnUSD_score.lastPriceInLoop()
```
Comment (dbrehmer) [inline rebalancing/rebalancing.py:261]: This is the same as the oracle price on line 194.
getPoolStats has everything you need to get the price on line 262, so you can save calls to the oracle and the DEX. Get the number of tokens in the pool here with getPoolStats, get the oracle price, pass them to the _calculate_tokens_to_retire method.
Also, you have missed the price of sICX in loop, which you need to get the true price to compare with the DEX price.

You do this calculation three times - `10 ** 36 // pool_price_dex`
Assign `dex_price = pool_stats['base'] * EXA // pool_stats['quote']`
    where, `EXA = 10 ** 18`

Try defining a constant `TOKENS = ["sICX", "bnUSD"]`
then,
```
def getRebalancingStatus(self) -> (bool, int, str):
    self.bnUSD_score = self.create_interface_score(self._bnUSD.get(), bnUSDTokenInterface)
    self.dex_score = self.create_interface_score(self._dex.get(), dexTokenInterface)
    self.sICX_score = self.create_interface_score(self._sicx.get(), sICXTokenInterface)
    price = self.bnUSD_score.lastPriceInLoop() * EXA // self.sICX_score.lastPriceInLoop()
    pool_stats = dex_score.getPoolStats(2)
    dex_price = pool_stats['base'] * EXA // pool_stats['quote']
    direction = price < dex_price
    diff = (dex_price - price) * EXA // price
    max_diff = self._price_threshold.get()
    if not -max_diff < diff < max_diff:
        return True, self._calculate_tokens_to_retire(price, pool_stats['base'], pool_stats['quote'], direction), TOKENS[direction]
    return False, 0, ""
```

File: rebalancing/rebalancing.py [hunk 2]
```diff
@@ -0,0 +1,311 @@
...
+           Calls the retireRedeem function of loans and retire the value of bnUSD.
+           Rebalances only if the rate of change in dex pool price and oracle price is greater than the threshold set.
+        """
+        self.sICX_score = self.create_interface_score(self._sicx.get(), sICXTokenInterface)
+        self.bnUSD_score = self.create_interface_score(self._bnUSD.get(), bnUSDTokenInterface)
+        rebalancing_status = self.getRebalancingStatus()
+        if rebalancing_status[0]:
+            if rebalancing_status[2] == "sICX":
+                sicx_to_retire = rebalancing_status[1]
+                sicx_in_contract = self.sICX_score.balanceOf(self.address)
```
Comment (dbrehmer) [inline rebalancing/rebalancing.py:290]: Right now there is no control over the tokenFallback so anyone could change the sicx_to_retire amount by adding more sicx. Probably best to make this a configurable parameter. Could retire the lesser of the parameter amount or the SCORE balance.

**Review by dbrehmer — CHANGES_REQUESTED** [review]
Lots of things to change. Please revise to have only the bnUSD retirement version enabled. We need to consider how to handle the other direction a little more.

**Comment by dbrehmer** [comment]
Let's have a call in the morning to discuss.

File: rebalancing/rebalancing.py [hunk 3]
```diff
@@ -0,0 +1,311 @@
...
+            # value = (self._sqrt(oracle_rate * sicx_supply * bnusd_supply) // 10 ** 9) - bnusd_supply
+            value = (self._sqrt((sicx_supply * bnusd_supply) // oracle_rate) * 10 ** 9) - bnusd_supply
+        return value
+
+    @external
+    @only_governance
+    def setPriceChangeThreshold(self, _value: int) -> None:
+        """
+        :param _value: threshold to set.
+        Sets the threshold .
```
[the remark this answers is not in the document]
Comment (Suyog007) [reply rebalancing/rebalancing.py:209]: yeah

File: rebalancing/rebalancing.py [hunk 4]
```diff
@@ -0,0 +1,311 @@
...
+
+    @external(readonly=True)
+    def getRebalancingStatus(self) -> list:
+        """
+        Checks the Rebalancing status of the pool.
+        """
+        rebalancing_direction = 0
+        self.bnUSD_score = self.create_interface_score(self._bnUSD.get(), bnUSDTokenInterface)
+        self.dex_score = self.create_interface_score(self._dex.get(), dexTokenInterface)
+        self.loans_score = self.create_interface_score(self._loans.get(), loansTokenInterface)
```
[the remark this answers is not in the document]
Comment (Suyog007) [reply rebalancing/rebalancing.py:260]: Loans score by defining here we can use it in rebalance function.We don't need to define separately in rebalance function.

File: rebalancing/rebalancing.py (hunk 2 above)
Comment (Suyog007) [reply rebalancing/rebalancing.py:290]: Do you mean setting a parameter by owner that sets 1000 value for now?

File: core_contracts/loans/loans/loans.py [hunk 5]
```diff
@@ -621,59 +630,68 @@ def returnAsset(self, _symbol: str, _value: int, _repay: bool = True) -> None:
...
+        asset.burnFrom(_from, _redeemed)
         for _ in range(min(batch_size, len(borrowers))):
             user_debt = borrowers.node_value(node_id)
             positions_dict[node_id] = user_debt
             total_batch_debt += user_debt
             borrowers.move_head_to_tail()
             node_id = borrowers.get_head_id()
         borrowers.serialize()
-
-        if POINTS * _redeemed > self._max_retire_percent.get() * total_batch_debt:
```
[the remark this answers is not in the document]
Comment (Suyog007) [reply core_contracts/loans/loans/loans.py:675]: If this reverts, then even the next time during retiring , same borrowers batch will be present right?

**Comment by BennyOptions** [comment]
This PR is for the rebalancing SCORE deployment only, please clean up

**Comment by dbrehmer** [comment]
Please remove all reverse rebalancing code from this branch and PR. It is too complicated for one PR.

File: core_contracts/loans/loans/loans.py [hunk 6]
```diff
@@ -512,9 +513,14 @@ def tokenFallback(self, _from: Address, _value: int, _data: bytes) -> None:
             d = json_loads(_data.decode("utf-8"))
         except BaseException as e:
             revert(f'{TAG}: Invalid data: {_data}, returning tokens. Exception: {e}')
-        if set(d.keys()) != {"_asset", "_amount"}:
+        if set(d.keys()) == {"_asset", "_amount"} or ('method' in set(d.keys()) and d["method"] == "_generate_bnUSD"):
```
[the remark this answers is not in the document]
Comment (Suyog007) [reply core_contracts/loans/loans/loans.py]: For now we will be requiring data in the forward case too. So to maintain consistency,it takes {"_asset": "bnUSD", "_amount": ""} as a data in bytes while sending sICX token to loans contract.

File: rebalancing/rebalancing.py (hunk 2 above)
Comment (Suyog007) [reply rebalancing/rebalancing.py]: Didn't quite get you.
But In rebalancing contract token fallback, we will revert back all the sICX sending transaction if it is not from the loans contract or owner. In that case the sicx_to_retire won't be changed.

File: rebalancing/rebalancing.py [hunk 7]
```diff
@@ -0,0 +1,311 @@
...
+    @external(readonly=True)
+    def getPriceChangeThreshold(self) -> int:
+        """
+        Returns the threshold value set by Governance contract.
+        """
+        return self._price_threshold.get()
+
+    @external
+    @only_governance
+    def setSicxReceivable(self, _value: int) -> None:
```
[the remark this answers is not in the document]
Comment (Suyog007) [reply rebalancing/rebalancing.py:185]: It's for the forward case only. This function ets the sICX amount to receive by rebalancing contract from the loans contract.

File: core_contracts/loans/loans/loans.py [hunk 8]
```diff
@@ -563,47 +569,50 @@ def depositAndBorrow(self, _asset: str = '', _amount: int = 0,
...
+        redeemed = _value
+        bad_debt = asset.bad_debt.get()
+        sicx: int = 0
+        if bad_debt > 0:
+            day, new_day = self.checkForNewDay()
+            self.checkDistributions(day, new_day)
+            bd_value = min(bad_debt, redeemed)
+            asset.burnFrom(_from, bd_value)
+            redeemed -= bd_value
+            sicx += self.bd_redeem(_from, asset, bd_value, sicx_rate, price)
```
[the remark this answers is not in the document]
Comment (Suyog007) [reply core_contracts/loans/loans/loans.py]: did the changes.You can view that from line 757.

File: core_contracts/loans/loans/loans.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (Suyog007) [reply core_contracts/loans/loans/loans.py:675]: Now it just changes the batch ,if it goes to the condition where the bnUSD we are retiring exceeds the minimum retire amount.

File: rebalancing/rebalancing.py [hunk 9]
```diff
@@ -0,0 +1,311 @@
...
+    def on_install(self) -> None:
+        super().on_install()
+        self._admin.set(self.msg.sender)
+
+    def on_update(self) -> None:
+        super().on_update()
+
+    @external
+    @only_admin
+    def setbnUSD(self, _address: Address) -> None:
```
[the remark this answers is not in the document]
Comment (Suyog007) [reply rebalancing/rebalancing.py]: Still working on that.

File: core_contracts/loans/utils/checks.py [hunk 10]
```diff
@@ -1,72 +1,81 @@
...
-	pass
+    pass
 
 
 class SenderNotGovernance(Exception):
-	pass
+    pass
+
+
+class SenderNotRebalance(Exception):
```
Comment (dbrehmer) [inline core_contracts/loans/utils/checks.py:20]: Need an only_rebalance method defined here to go with this.

File: core_contracts/rebalancing/rebalancing.py [hunk 11]
```diff
@@ -0,0 +1,222 @@
...
+        return y
+
+    def _calculate_tokens_to_retire(self, price: int, base_supply: int, quote_supply: int) -> int:
+        """
+        :param price: Oracle price.
+        :param base_supply: base token supply.
+        :param quote_supply: quote token supply.
+        Returns the amount of sICX required for rebalancing the price.
+        """
+        value = (self._sqrt(price * base_supply * quote_supply) // 10 ** 9) - base_supply
```
Comment (dbrehmer) [inline core_contracts/rebalancing/rebalancing.py:150]: Can bring the 10 ** 9 into the sqrt and use the EXA constant defined above.

**Review by dbrehmer — APPROVED** [review]
There are a bunch of needed changes, but it will work okay as is, except the need for a guard for the retireRedeem method.
In any case, I am going to approve now and we can make the final changes in an additional PR so this one does not keep expanding.

**Review by dbrehmer — APPROVED** [review]
Just merged in Development

**Pull request merged**